selfservice/strategy/oidc/strategy.go

```diff
@@ -589,3 +591,38 @@ func (s *Strategy) CompletedAuthenticationMethod(ctx context.Context) session.Au
 		AAL:    identity.AuthenticatorAssuranceLevel1,
 	}
 }
+
+func (s *Strategy) processIDToken(w http.ResponseWriter, r *http.Request, provider Provider, idToken, idTokenNonce string) (*Claims, error) {
+	verifier, ok := provider.(IDTokenVerifier)
+	if !ok {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("The provider %s does not support id_token verification", provider.Config().Provider))
+	}
+	claims, err := verifier.Verify(r.Context(), idToken)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Could not verify id_token").WithError(err.Error()))
+	}
+
+	if err := claims.Validate(); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("The id_token claims were invalid").WithError(err.Error()))
+	}
+
+	// Not all providers support nonce, so we only check if the provider supports it.
+	if verifier.NonceSupported(claims) {
+		if idTokenNonce == "" {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("No nonce was provided but is required by the provider"))
+		}
+
+		if claims.Nonce == "" {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("No nonce was included in the id_token but is required by the provider"))
+		}
+		sh := sha256.New()
+		sh.Write([]byte(idTokenNonce))
+		hashedNonce := hex.EncodeToString(sh.Sum(nil))
+
+		if hashedNonce != claims.Nonce {
+			return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("The supplied nonce does not match the nonce from the id_token"))
+		}
+	}
```

Comment:
Why is this being hex encoded? From the OIDC specification:

> **nonce:** String value used to associate a Client session with an ID Token, and to mitigate replay attacks. The value is passed through unmodified from the Authentication Request to the ID Token. If present in the ID Token, Clients MUST verify that the nonce Claim Value is equal to the value of the nonce parameter sent in the Authentication Request. If present in the Authentication Request, Authorization Servers MUST include a nonce Claim in the ID Token with the Claim Value being the nonce value sent in the Authentication Request. Authorization **Servers SHOULD perform no other processing on nonce values used. The nonce value is a case sensitive string.**

https://openid.net/specs/openid-connect-core-1_0.html

https://github.com/coreos/go-oidc/blob/f0117a5f178cb36710dde42ede2154071b20ee94/example/idtoken/app.go#L109-L117

If this check is Apple-specific, it should live in the apple strategy IMO.

Reply:
Fixed. Thanks for the sources.